lib/pipekit/request.rb

```diff
@@ -32,48 +32,66 @@ def initialize(resource)
     #   search_by_field(field: :cohort, value: 119)
     #   search_by_field(field: :github_username, value: "octocat")
     #
-    # Returns an array of Hashes or nil.
+    # Returns an array of Response objects or throws a ResourceNotFoundError if
+    # it couldn't find anything.
     def search_by_field(field:, value:)
       query = {field_type: "#{resource}Field",
                field_key: Config.field(resource, field),
                return_item_ids: true,
                term: value
       }
 
-      result_from self.class.get("/searchResults/field", options(query: query))
+      response_from self.class.get("/searchResults/field", options(query: query))
     end
 
-    def get(suffix = nil, query = {})
-      result_from self.class.get(uri(suffix), options(query: query))
+    # Public: Pipedrive GET API call - does a GET request to the Pipedrive API
+    # based on the resource passed in the initialiser
+    #
+    # id - If the resource being searched for has an id
+    # query - An optional query string
+    # start - The offset with which to start the query
+    #
+    # As long as "request_all_pages" is not set to false in the config this will
+    # recursively call `#get` until all the pages of the request have been
+    # fetched from pipedrive
+    # Pipedrive until everything available has been received
+    def get(id = nil, query = {})
+      _get(id, query, get_request(id, query))
     end
 
     def put(id, body)
-      result_from self.class.put(uri(id), options(body: body))
+      response_from self.class.put(uri(id), options(body: body))
     end
 
     def post(body)
-      result_from self.class.post(uri, options(body: body))
+      response_from self.class.post(uri, options(body: body))
     end
 
     private
 
     attr_reader :resource
 
-    def uri(id = nil)
-      "/#{resource}s/#{id}".chomp("/")
+    def _get(id, query, result)
+      return result.response(resource) unless result.fetch_next_request?
+      _get(id, query, result + get_request(id, query, result.next_start))
     end
 
-    def result_from(response)
-      data = response["data"]
-      success = response["success"]
+    def get_request(id, query, start = 0)
+      response = self.class.get(uri(id), options(query: {limit: pagination_limit, start: start}.merge(query)))
+      Result.new(response)
```

Comment:
That's cool - I like using the Result object to understand when all the data has been gathered.